bw2data/revisions.py

```diff
@@ -166,10 +173,15 @@ def generate(
         if operation is not None:
             return getattr(cls, operation)(old, new)
 
-        if old is None and new is None:
+        if old is not None:
+            obj_id = old.id
+            obj_type = old.__class__
+        elif new is not None:
+            obj_id = new.id
+            obj_type = new.__class__
+        else:
```

Comment:
No amount of `if`/`assert` massaging was enough to make `mypy` happy about these mutually-exclusive conditions, so I've reworked the code here a bit. The logic should remain the same.